README
AnyEvent-Redis

AnyEvent::Redis is an event-driven (asynchronous) client for the Redis
key-value (NoSQL) database server.  Both standard query and subscriber modes
are available.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc AnyEvent::Redis
    perldoc AnyEvent::Redis::Subscriber

SOURCE CODE

The official Git repository for AnyEvent::Redis is located at
https://github.com/otterley/AnyEvent-Redis.

SEE ALSO

Redis <http://redis.io>

LICENSE AND COPYRIGHT

Copyright (C) 2010 Michael S. Fischer

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for AnyEvent-Redis

0.1     Mon Dec  6 00:34:17 PST 2010  Michael S. Fischer <michael+cpan@dynamine.net>
        Initial release.



lib/AnyEvent/Redis/Protocol.pm
package AnyEvent::Redis::Protocol;

use common::sense;

=head1 NAME

AnyEvent::Redis::Protocol - generates and parses Redis messages

=head1 DESCRIPTION

This package should not be directly used.  It provides AnyEvent write and read
types capable of generating Redis requests and parsing replies.

=head1 SEE ALSO

L<AnyEvent::Handle>,
Redis Protocol Specification L<http://code.google.com/p/redis/wiki/ProtocolSpecification>

=cut

sub anyevent_write_type {
    my ($handle, @args) = @_;

    use bytes;
    return "*" . scalar @args . "\015\012" .
          join("\015\012", map { '$' . length($_) . "\015\012" . $_ } @args) .
          "\015\012";
}

sub anyevent_read_type {
    my ($handle, $cb) = @_;
    return sub {
        $handle->push_read(line => sub {
            my $line = $_[1];
            if ($line =~ /^\*(-?\d+)/) {
                my $nexpected = $1;
                if ($nexpected == -1) {
                    $cb->($handle, undef, undef);
                } else {
                    my $values = [];
                    for (my $i = 0; $i < $nexpected; $i++) {
                        $handle->push_read('AnyEvent::Redis::Protocol' => sub {
                                my ($handle, $errmsg, $value) = @_;
                                push @$values, $value;
                                if (scalar @$values == $nexpected) {
                                    $cb->($handle, undef, @$values);
                                }
                        });
                    }
                }
            }
            if ($line =~ /^[\+:](.*)/) {
                # Single line/integer reply
                $cb->($handle, undef, $1);
            } elsif ($line =~ /^-(.*)/) {
                # Single-line error reply
                $cb->($handle, $1);
            } elsif ($line =~ /^\$(-?\d+)/) {
                # Bulk reply
                my $length = $1;
                if ($length == -1) {
                    $cb->($handle, undef, undef);
                } else {
                    # We need to read 2 bytes more than the length (stupid 
                    # CRLF framing).  Then we need to discard them.
                    $handle->unshift_read(chunk => $length + 2, sub {
                        use bytes;
                        my $data = $_[1];
                        $cb->($handle, undef, substr($data, 0, length($data) - 2));
                    });
                }
            }
        });
        1;
    };
}

=head1 AUTHOR

Michael S. Fischer <michael+cpan@dynamine.net>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 Michael S. Fischer.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;

__END__

# vim:syn=perl:ts=4:sw=4:et:ai


lib/AnyEvent/Redis/Subscriber.pm
package AnyEvent::Redis::Subscriber;

use 5.008;
use common::sense;

use base 'Object::Event';
use Carp qw(croak);
use AnyEvent::Handle;
use AnyEvent::Redis::Protocol;

=head1 NAME

AnyEvent::Redis::Subscriber - event-driven Redis subscriber client

=head1 SYNOPSIS

 use AnyEvent::Redis::Subscriber;
 my $subscriber = AnyEvent::Redis::Subscriber->subscribe(
     host => 'redis', 
     channel => 'z'
 );
 $subscriber->reg_cb(message => sub {
     my ($subscriber, $message) = @_;
     # ...
 };

=head1 DESCRIPTION

This module is an AnyEvent user; you must use and run a supported event
loop.

AnyEvent::Redis::Subscriber is an event-driven (asynchronous) subscriber for
the Redis publish/subscribe system.  

The main idea is that other clients (such as AnyEvent::Redis) publish messages
to a specified channel (or class, in Redis-speak), which are then received by
each client that has subscribed to the channel.

Unlike some other publish/subscribe mechanisms which are producer-consumer
based, Redis' publish/subscribe mechanism is broadcast-based: every subscriber
to a channel receives every message sent to that channel; and messages are not
queued for future delivery if there are no subscribers for that channel.

=head2 Subscribing to a channel

To subscribe to a channel, call the subscribe() method:

 my $subscriber = AnyEvent::Redis::Subscriber->subscribe(
    host => $host,
    channel => $channel,
    ...
 );

This immediately establishes a connection to the Redis server at the given host
and subscribes to the specified channel. 

Pattern-based subscriptions are also possible.  Instead of specifying a
C<channel>, a C<pattern> may be specified consisting of a glob pattern that
matches channel names.  For example, if C<chan_*> is specified, then the client
will receive every message sent to "chan_1", "chan_2", "chan_zebra", and so
forth while connected to the server:

 my $subscriber = AnyEvent::Redis::Subscriber->subscribe(
    host => $host,
    pattern => $pattern,
    ...
 );

Optional arguments include:

=over 

=item port => $port

Connect to the Redis server at L<$port>.  If undefined, the default port (6379)
is used.

=item auth => $password

Authenticate to the server with the given password.

=back

=head2 Receiving messages

To receive messages, register a callback for the C<message> event.  When a
message is received, the callback will be fired with the message and the
channel it was received on:

 $subscriber->reg_cb(message => sub {
     my ($subscriber, $channel, $message) = @_;
     print "Received message $message on channel $channel";
 });

=head2 Dealing with errors

To be notified of errors (which can occur at connect time or thereafter), 
register a callback as follows:

 $subscriber->reg_cb(error => sub {
    my ($subscriber, $errmsg) = @_;
    # ...
 });

B<WARNING:> As with all AnyEvent modules, B<you must not die() in a callback!> 

=head2 Unsubscribing

To unsubscribe, just undef the subscriber handle.  Reusing handles for
subsequent subscriptions is not supported.  Even though the wire protocol
technically supports it, trust me, it's better this way :-).

=cut

sub subscribe {
    my ($class, %args) = @_;
    
    $args{host} or croak "Missing host";
    $args{channel} or $args{pattern} or croak "Missing channel/pattern";
    $args{channel} and $args{pattern} and croak "Both pattern and channel specified";
    $args{port} ||= 6379;

    my $self = $class->SUPER::new(%args);
    bless $self, $class;

    $self->{handle} = AnyEvent::Handle->new(
        connect => [ $self->{host}, $self->{port} ],
        keepalive  => 1,
        no_delay   => 1,
        on_connect => sub { 
            my ($handle, $host, $port) = @_;
            if ($self->{auth}) {
                $self->{handle}->push_write('AnyEvent::Redis::Protocol', 
                                            AUTH => $self->{auth});
                $self->{handle}->push_read('AnyEvent::Redis::Protocol' => sub {
                        my ($handle, $errmsg) = @_;
                        if ($errmsg) {
                            $self->{handle}->destroy;
                            $self->event('error', $errmsg);
                        } else {
                            $self->_subscribe;
                        }
                });
            } else {
                $self->_subscribe;
            }
        },
        on_connect_error => sub { 
            $self->event('error', $_[1]);
        },
        on_error => sub { 
            $self->{handle}->destroy; 
            $self->event('error', $_[2]);
        }, 
    );

    return $self;
}

sub _subscribe {
    my $self = shift;
    if ($self->{pattern}) {
        $self->{handle}->push_write('AnyEvent::Redis::Protocol', 
                                    PSUBSCRIBE => $self->{pattern});
    } else {
        $self->{handle}->push_write('AnyEvent::Redis::Protocol', 
                                    SUBSCRIBE => $self->{channel});
    }
    $self->{handle}->push_read('AnyEvent::Redis::Protocol' => sub {
            my ($handle, $errmsg, @values) = @_;
            if ($errmsg) { 
                $self->event('error', $errmsg);
            } else {
                if (($self->{channel} && $values[0] eq 'subscribe' && $values[1] eq $self->{channel})
                    || ($self->{pattern} && $values[0] eq 'psubscribe' && $values[1] eq $self->{pattern})) {
                    $self->{handle}->on_read(sub {
                            $self->{handle}->push_read('AnyEvent::Redis::Protocol' => sub {
                                    if ($_[2] eq 'message') {
                                        $self->event('message', $_[3], $_[4]);
                                    } elsif ($_[2] eq 'pmessage') {
                                        $self->event('message', $_[4], $_[5]);
                                    }
                            });
                    });
                } else {
                    $self->event('error', 'Unexpected response to [P]SUBSCRIBE command');
                }
            }
    });
}

=head1 SEE ALSO

Redis PublishSubscribe L<http://code.google.com/p/redis/wiki/PublishSubscribe>

=head1 AUTHOR

Michael S. Fischer <michael+cpan@dynamine.net>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 Michael S. Fischer

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;

__END__

# vim:syn=perl:ts=4:sw=4:et:ai


lib/AnyEvent/Redis.pm
package AnyEvent::Redis;

use 5.008;
use common::sense;

use Carp qw(croak);
use AnyEvent::Handle;
use AnyEvent::Redis::Protocol;

our $VERSION = 0.1;

=head1 NAME

AnyEvent::Redis - an event-driven Redis client

=head1 SYNOPSIS

 use AnyEvent::Redis;
 my $redis = AnyEvent::Redis->connect(
     host => 'redis',
     on_connect => $cb, 
     on_connect_error => $errcb,
     on_error => $errcb,
 );

 # Generic query method
 $redis->query('SET', $key, $value, $cb);

 # Autoloaded query method
 $redis->set($key, $value, $cb);

=head1 DESCRIPTION

This module is an AnyEvent user; you must use and run a supported event
loop.

AnyEvent::Redis is an event-driven (asynchronous) client for the Redis
key-value (NoSQL) database server.  Every operation is supported, B<except>
subscription to Redis "classes" (i.e. channels), which is supported by
AnyEvent::Redis::Subscriber.  However, this module may be used to publish to
channels.

=head2 Establishing a connection

To connect to the Redis server, use the connect() method:

 my $redis = AnyEvent::Redis->connect(host => <host>, ...);

The C<host> argument is required.

Optional (but recommended) arguments include:

=over

=item port => $port

Connect to the server on the specified port number.  (If not specified,
the default port will be used.)

=item auth => $password

Authenticate to the server with the given password.

=item on_connect => $cb->($conn, $host, $port)

Specifies a callback to be executed upon a successful connection.  The
connection object and the actual peer host and port number will be passed as
arguments to the callback.

=item on_connect_error => $cb->($conn, $errmsg)

Specifies a callback to be executed if the connection failed (or authentication
failed).  The connection object and the error message will be passed as
arguments to the callback.  The connection is not reusable.

=item on_error => $cb->($conn, $errmsg)

Specifies a callback to be executed if an I/O error occurs (e.g. connection
reset by peer).  The connection object and the error message will be passed as
arguments to the callback.  The connection is not reusable.

=back

=cut

sub connect {
    my ($class, %args) = @_;
    
    $args{host} or croak "Missing host";
    $args{port} ||= 6379;

    my $self = { %args };
    bless $self, $class;

    $self->{handle} = AnyEvent::Handle->new(
        connect => [ $self->{host}, $self->{port} ],
        keepalive  => 1,
        no_delay   => 1,
        on_connect => sub { 
            my ($handle, $host, $port) = @_;
            if ($self->{auth}) {
                $self->{handle}->push_write('AnyEvent::Redis::Protocol', 
                                            AUTH => $self->{auth});
                $self->{handle}->push_read('AnyEvent::Redis::Protocol' => sub {
                        my ($handle, $ok) = @_;
                        if ($ok) {
                            $self->{on_connect}->($self, $host, $port) 
                                if $self->{on_connect};
                        } else {
                            $self->{handle}->destroy;
                            $self->{on_connect_error}->($self, $_[2])
                                if $self->{on_connect_error};
                        }
                });
            } else {
                $self->{on_connect}->($self, $host, $port) 
                    if $self->{on_connect};
            }
        },
        on_connect_error => sub { 
            $self->{on_connect_error}->($self, $_[1]) 
                if $self->{on_connect_error};
        },
        on_error   => sub { 
            $self->{handle}->destroy; 
            $self->{on_error}->($self, $_[2]) if $self->{on_error};
        }, 
    );

    return $self;
}

=head2 Queries and responses

After you have successfully connected to the server, you can issue 
queries to it.  (You'll want to do this in the C<on_connect> callback
handler fired by connect(), above.)  

To issue a query, use the query() method:

  $redis->query(@args, $cb->($errmsg, @data)));

The initial list of arguments to query() comprise the actual command.
(See the command reference
L<http://code.google.com/p/redis/wiki/CommandReference> for a list of available
commands.)

The final argument to query() specifies a callback (code reference) that will
be fired when a response to the query is received.  It will be called with the
error message (which will be C<undef> if there was no error) as the first
argument; the remaining arguments will be the values (if any) that comprise the
response.  

=cut

sub query {
    my ($self, @args) = @_;

    my $cb = pop @args;
    ref($cb) eq 'CODE' or croak 'Missing callback';
    scalar @args or croak "Missing args";

    $args[0] !~ /^P?(?:UN)?SUBSCRIBE/ 
        or croak 'Subscriptions not supported; use AnyEvent::Redis::Subscriber instead';

    $self->{handle}->push_write('AnyEvent::Redis::Protocol', @args);
    $self->{handle}->push_read('AnyEvent::Redis::Protocol' => sub {
        my ($handle, $errmsg, @values) = @_;
        $cb->($errmsg, @values);
    });
}

=pod

Alternatively, you may invoke Redis commands as methods, e.g.: 

 $redis->set(key1 => $val1, sub { ... });
 $redis->get(key1 => sub { ... });

=cut

sub AUTOLOAD {
    use vars '$AUTOLOAD';
    my $op = $AUTOLOAD;
    $op =~ s/.*:://;
    my ($self, @args) = @_;
    $self->query(uc($op), @args);
}

sub DESTROY {
    # Don't delete this, or AUTOLOAD will be invoked on DESTROY.
}

=head1 SEE ALSO

Redis Command Reference L<http://code.google.com/p/redis/wiki/CommandReference>

=head1 AUTHOR

Michael S. Fischer <michael+cpan@dynamine.net>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 Michael S. Fischer.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;

__END__

# vim:syn=perl:ts=4:sw=4:et:ai
